extensions/ql-vscode/src/data-extensions-editor/auto-model.ts

```diff
@@ -0,0 +1,117 @@
+import { ExternalApiUsage } from "./external-api-usage";
+import { ModeledMethod, ModeledMethodType } from "./modeled-method";
+import {
+  Classification,
+  ClassificationType,
+  Method,
+  ModelRequest,
+} from "./auto-model-api";
+
+export function createAutoModelRequest(
+  language: string,
+  externalApiUsages: ExternalApiUsage[],
+  modeledMethods: Record<string, ModeledMethod>,
+): ModelRequest {
+  const request: ModelRequest = {
+    language,
+    samples: [],
+    candidates: [],
+  };
+
+  // Sort by number of usages so we always send the most used methods first
+  externalApiUsages = [...externalApiUsages];
+  externalApiUsages.sort((a, b) => b.usages.length - a.usages.length);
+
+  for (const externalApiUsage of externalApiUsages) {
+    const modeledMethod: ModeledMethod = modeledMethods[
+      externalApiUsage.signature
+    ] ?? {
+      type: "none",
+    };
+
+    const numberOfArguments =
+      externalApiUsage.methodParameters === "()"
+        ? 0
+        : externalApiUsage.methodParameters.split(",").length;
+
+    for (
+      let argumentIndex = 0;
+      argumentIndex < numberOfArguments;
+      argumentIndex++
+    ) {
+      const method: Method = {
+        package: externalApiUsage.packageName,
+        type: externalApiUsage.typeName,
+        name: externalApiUsage.methodName,
+        signature: externalApiUsage.methodParameters,
+        classification:
+          modeledMethod.type === "none"
+            ? undefined
+            : toMethodClassification(modeledMethod),
+        usages: externalApiUsage.usages.map((usage) => usage.label),
+        input: `Argument[${argumentIndex}]`,
+      };
+
+      if (method.usages.length > 10) {
+        method.usages = method.usages.slice(0, 10);
+      }
+
+      if (modeledMethod.type === "none") {
+        request.candidates.push(method);
+      } else {
+        request.samples.push(method);
+      }
+    }
+  }
+
```

Comment:
Now that we've created a `Method` for each argument, does it make sense to shuffle the array before slicing, to avoid ending up with the same method multiple times?

Reply:
I think we'll lose the sorting by number of usages in that case. I think we still do want that prioritization.

This also made me think of the processing of the `predicted` methods. If we send the same method multiple times, we'll also get multiple results. Right now the editor only supports 1 classification per method, so we would get a "random" classification from the supplied classifications. I'm not sure how to best deal with that.

Reply:
I think there are general questions about how to deal with the response, given we classify a single argument.

We return from the LLM either "sink" or "not sink" for a specific argument, so unless the response is sink we cannot actually fill in anything. setting it to neutral would be wrong as that specific argument could be either a source or a summary, and other non checked arguments could still be sinks. I think this is a broader discussion we should record, but dont have to solve now.

I propose we do the following now:
- If there is a single argument of the method that is a sink, then pick the lowest argument index and use that as the result.
- If all arguments are non-sink then model this as neutral.